app/queues/reconciler/reconciler-worker.server.ts

```diff
@@ -0,0 +1,154 @@
+import { Worker, Queue } from 'bullmq';
+import { redis } from '~/lib/redis.server';
+import { executeChangeSet } from '~/lib/dns.server';
+import logger from '~/lib/logger.server';
+import DnsDbCompareStructureGenerator from './DnsDbCompareStructureGenerator.server';
+import Route53CompareStructureGenerator from './Route53CompareStructureGenerator.server';
+import {
+  createRemovedChangeSetFromCompareStructures,
+  createUpsertedChangeSetFromCompareStructures,
+} from './createChangeSetFromCompareStructures.server';
+import { getIsReconciliationNeeded, setIsReconciliationNeeded } from '~/models/system-state.server';
+import type { Change } from '@aws-sdk/client-route-53';
+
+// S3 limit for a ChangeSet
+const CHANGE_SET_MAX_SIZE = 1000;
+const reconcilerQueueName = 'reconciler';
+
+// Queue initialization
+export const reconcilerQueue = new Queue(reconcilerQueueName, {
+  connection: redis,
+});
+
+const createChangeSet = async (): Promise<Change[]> => {
+  const [dbStructure, route53Structure] = await Promise.all([
+    new DnsDbCompareStructureGenerator().generate(),
+    new Route53CompareStructureGenerator().generate(),
+  ]);
+
+  const changeSet = [
+    ...createRemovedChangeSetFromCompareStructures({ dbStructure, route53Structure }),
+    ...createUpsertedChangeSetFromCompareStructures({ dbStructure, route53Structure }),
+  ];
+
+  return changeSet;
+};
+
+/**
+ * NORMAL MODE
+ *
+ * Execute the complete changeSet at once
+ */
+const pushChangesBulk = async (changeSet: Change[]): Promise<boolean> => {
+  const recordSetsToPush = Math.min(CHANGE_SET_MAX_SIZE, changeSet.length);
+
+  logger.debug(
+    `Reconciler NORMAL MODE - Reconciler intends to push the following ${recordSetsToPush} changes`,
+    {
+      changeSet,
+    }
+  );
+
+  await executeChangeSet(changeSet.slice(0, CHANGE_SET_MAX_SIZE));
+
+  // Return boolean => Is additional reconciliation needed
+  return changeSet.length > CHANGE_SET_MAX_SIZE;
+};
+
+/**
+ * LIMP MODE
+ *
+ * Try each change in the set one by one, isolate the offending one
+ */
+const pushChangesIndividually = async (changeSet: Change[]) => {
+  for (const change of changeSet) {
+    try {
+      logger.debug(`Reconciler LIMP MODE - Reconciler intends to push the following change`, {
+        change,
+      });
+
+      await executeChangeSet([change]);
+    } catch (error) {
+      logger.error(`Reconciler LIMP MODE - the following single change failed`, {
+        change,
+        error,
+      });
+    }
+  }
+};
+
+const reconcilerWorker = new Worker(
+  reconcilerQueueName,
+  async () => {
+    /**
+     * When a BullMQ worker is added, it will behave as single
```

Comment:
This is a great discussion of what's happening here, thanks for adding it.

Reply:
Absolutely!